samcli/lib/config/file_manager.py

```diff
@@ -0,0 +1,187 @@
+"""
+Class to represent the parsing of different file types into Python objects.
+"""
+
+
+import logging
+from abc import ABC, abstractmethod
+from pathlib import Path
+from typing import Any
+
+import tomlkit
+
+from samcli.lib.config.exceptions import FileParseException
+
+LOG = logging.getLogger(__name__)
+
+
+class FileManager(ABC):
+    """
+    Abstract class to be overridden by file managers for specific file extensions.
+    """
+
+    @staticmethod
+    @abstractmethod
+    def read(filepath: Path) -> Any:
+        """
+        Read a file at a given path.
+
+        Parameters
+        ----------
+        filepath: Path
+            The Path object that points to the file to be read.
+
+        Returns
+        -------
+        Any
+            A dictionary-like representation of the contents at the filepath location, along with a specialized
+            representation of the file that was read, if there is a specialization of it.
+        """
+        raise NotImplementedError("Read method not implemented.")
+
+    @staticmethod
+    @abstractmethod
+    def write(document: dict, filepath: Path):
+        """
+        Write a dictionary to a given file.
+
+        Parameters
+        ----------
+        document: dict
+            The object to write.
+        filepath: Path
+            The final location for the document to be written.
+        """
+        raise NotImplementedError("Write method not implemented.")
+
+    @staticmethod
+    @abstractmethod
+    def write_document(document: Any, filepath: Path):
```

Comment:
We could change them to `write_dict()` and `write_document()` if that adds more clarity. I'm thinking that `write_dict()` will write a generic dictionary to its respective document type, while `write_document()` writes the file document object, which may have its own special formatting. In the case of TOML, it would take in a `tomlkit.TOMLDocument` and write it, which would have support for comments.